xdsl/dialects/riscv.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import Annotated
+
+from xdsl.ir import (
+    Dialect,
+    Operation,
+    SSAValue,
+    Data,
+    OpResult,
+    TypeAttribute,
+)
+
+from xdsl.irdl import (
+    IRDLOperation,
+    irdl_op_definition,
+    irdl_attr_definition,
+    Operand,
+)
+
+from xdsl.parser import BaseParser
+from xdsl.printer import Printer
+
+
+@dataclass(frozen=True)
+class Register:
+    """
+    A RISC-V register.
+    """
+
+
+@irdl_attr_definition
+class RegisterType(Data[Register], TypeAttribute):
+    """
+    A RISC-V register type.
+    """
+
+    name = "riscv.reg"
+
+    @staticmethod
+    def parse_parameter(parser: BaseParser) -> Register:
+        return Register()
+
+    def print_parameter(self, printer: Printer) -> None:
+        pass
+
+
+class Riscv1Rd2RsOperation(IRDLOperation, ABC):
```

Comment:
Can this name be something more descriptive than just and encoding of the strucutre? This is essentially a `BinOpOnRegisters` or soemthing like that, right?

Reply:
Or a `RegisterBinOp` or something

Reply:
There are the classes that will eventually exist and here is a link to the RISCV spec (https://riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf). Maybe that helps to find better class names.

```
class Riscv1OffOperation(Operation):
class Riscv1Rd1ImmOperation(Operation):
class Riscv1Rd1OffOperation(Operation):
class Riscv1Rd1Rs1ImmOperation(Operation):
class Riscv1Rd1RsOperation(Operation):
class Riscv1Rd2RsOperation(Operation):
class Riscv1Rs1OffOperation(Operation):
class Riscv1Rs1Rt1OffOperation(Operation):
class Riscv2Rs1ImmOperation(Operation):
class Riscv2Rs1OffOperation(Operation):
class RiscvNoParamsOperation(Operation):
```

Reply:
I'm now thinking that it might be better private, the abstract class names are an implementation detail of how the inits and assembly printing code can be simplified. Would be interested in suggestions for better naming.

Reply:
This is also extensible if someone wants to extend the spec later on

Reply:
> So as per spec, these are the possible encodings (d = dest register, s = source register, i = immediate):
> 
> * `dss` - R type
> * `dsi` - I type
> * `ssi` - S/B type
> * `di` - U/J type
> * `dsss` - R4 type
> 
> I would recommend class names like `RiscvDSS`, or `RiscvRdRsImm`, if we want to be more explicit.

If we do this, it should be part of the documentation of the RISCV file, IMO. I think these names are besser, they are more concise and convey the same info.